### PR TITLE
Support custom terrain providers

### PR DESCRIPTION
**Pull request by pmconne**
Fixes https://github.com/iTwin/itwinjs-backlog/issues/66
Fixes https://github.com/iTwin/roadmap/issues/5

Primary API is in TerrainMeshProvider.ts.
Example custom terrain provider in display-test-app's BingTerrainProvider.ts. We can convert it to a sample/sandbox after 3.4 release.
The rest of the changes are mostly about (1) adjusting the existing providers to use the new APIs and (2) promoting to beta or public the minimal set of existing APIs required to implement a provider and documenting them.

### CODE REVIEW

**Comment by mgooding** [comment]
Still working my way through (there's a lot here) - but the API looks clean and reasonable to me. Looks like I'd have an easy time with it as a third-party developer.

**Comment by markschlosseratbentley** [comment]
> Still working my way through (there's a lot here) - but the API looks clean and reasonable to me. Looks like I'd have an easy time with it as a third-party developer.

Agreed.

File: core/frontend/src/tile/map/TerrainProvider.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+/** A registry of [[TerrainProvider]]s identified by their unique names. The registry can be accessed via [[IModelApp.terrainProviderRegistry]].
+ * It always includes the built-in provider named "CesiumWorldTerrain", which obtains terrain meshes from [Cesium ION](https://cesium.com/platform/cesium-ion/content/cesium-world-terrain/). That provider requires a valid [[TileAdmin.Props.cesiumIonKey]] to be supplied to [[IModelApp.startup]].
+ * Any number of additional providers can be [[register]]ed.
+ *
+ * When terrain is enabled for a [[Viewport]], the display system will attempt to look up the [[TerrainProvider]] corresponding to the [TerrainSettings.providerName]($common) specified by the [[Viewport]]'s [DisplayStyleSettings]($common). If a provider by that name is registered, it will be used to obtain terrain meshes; otherwise, the display system will produce flat terrain meshes.
+ * @beta
+ */
+export class TerrainProviderRegistry {
```
Comment (pmconne) [inline core/frontend/src/tile/map/TerrainProvider.ts:28]: @calebmshafer during roadmap planning you had an item entitled "provide a more clear and consise terrain/map layers API" having to do with some concerns about how map layer formats are registered. Do you have any problems with the API I've added for registering terrain providers?

File: core/bentley/src/TypedArrayBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+ * `TypedArrayBuilder` manages a `TypedArray`, permitting you to [[push]] and [[append]] elements to it. It exposes two "size" properties":
+ *  - [[capacity]], the number of elements it has currently allocated space for - i.e., the length of the underlying TypedArray; and
+ *  - [[length]], the number of elements that have so far been added to it, which is never bigger than [[capacity]].
+ * When [[capacity]] is exceeded, a new, bigger TypedArray is allocated and the contents of the previous array are copied over to it.
+ *
+ * Once you've finished adding elements, you can obtain the finished `TypedArray` via [[toTypedArray]].
+ * @see [[Uint8ArrayBuilder]], [[Uint16ArrayBuilder]], and [[Uint32ArrayBuilder]] to build specific types of arrays.
+ * @public
+ */
+export class TypedArrayBuilder<T extends Uint8Array | Uint16Array | Uint32Array> {
```
Comment (kabentley) [inline core/bentley/src/TypedArrayBuilder.ts:44]: Nice class. No tests?

Might be a nice enhancement to allow a "growthFunction" to have more complex growth policies. A future PR.

**Review by kabentley — APPROVED** [review]

File: core/bentley/src/TypedArrayBuilder.ts (hunk 2 above)
Comment (pmconne) [reply core/bentley/src/TypedArrayBuilder.ts:44]: It is indirectly tested by the VertexTableSplitter tests. Are you asking me to add some tests?

Comment (kabentley) [reply core/bentley/src/TypedArrayBuilder.ts:44]: No, i realize it's tested by your actual use of it in your tests and you want to complete this PR. I just think it's likely that we'll start using this often now that it exists, and tests often serve as examples.

**Review by markschlosseratbentley — APPROVED** [review]

**Review by calebmshafer — APPROVED** [review]
LGTM, reviewed the changes in iModelApp and doc updates.

**Pull request merged**